javascript/ql/lib/semmle/javascript/frameworks/data/internal/Shared.qll

```diff
@@ -0,0 +1,587 @@
+API::Node getNodeFromPath(string package, string type, AccessPath path) {
+  result = getNodeFromPath(package, type, path, path.getNumToken())
+}
+
+/**
+ * Gets an invocation identified by the given `(package, type, path)` tuple.
+ *
+ * Unlike `getNodeFromPath`, the `path` may end with one or more call-site filters.
+ */
+API::InvokeNode getInvocationFromPath(string package, string type, AccessPath path, int n) {
+  result = getNodeFromPath(package, type, path, n).getAnInvocation()
+  or
+  result = getInvocationFromPath(package, type, path, n - 1) and
+  invocationMatchesCallSiteFilter(result, path.getToken(n - 1))
+}
+
+/** Gets an invocation identified by the given `(package, type, path)` tuple. */
+API::InvokeNode getInvocationFromPath(string package, string type, AccessPath path) {
+  result = getInvocationFromPath(package, type, path, path.getNumToken())
+}
+
+/**
+ * Holds if a summary edge with the given `input, output, kind` columns have a `package, type, path` tuple
+ * that resolves to `baseNode`.
+ */
+private predicate resolvedSummaryBase(
+  API::InvokeNode baseNode, AccessPath input, AccessPath output, string kind
+) {
+  exists(string package, string type, AccessPath path |
+    summaryModel(package, type, path, input, output, kind) and
+    baseNode = getInvocationFromPath(package, type, path)
+  )
+}
+
+/**
+ * Holds if `path` is an input or output spec for a summary with the given `base` node.
+ */
+pragma[nomagic]
+private predicate relevantInputOutputPath(API::InvokeNode base, AccessPath path) {
+  resolvedSummaryBase(base, path, _, _)
+  or
+  resolvedSummaryBase(base, _, path, _)
+}
+
+/**
+ * Gets the API node for the first `n` tokens of the given input/output path, evaluated relative to `baseNode`.
+ */
+private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path, int n) {
+  relevantInputOutputPath(baseNode, path) and
+  (
+    n = 1 and
+    result = getSuccessorFromInvoke(baseNode, path.getToken(0))
+    or
+    result =
+      getSuccessorFromNode(getNodeFromInputOutputPath(baseNode, path, n - 1), path.getToken(n - 1))
+  )
+}
+
+/**
+ * Gets the API node for the given input/output path, evaluated relative to `baseNode`.
+ */
+private API::Node getNodeFromInputOutputPath(API::InvokeNode baseNode, AccessPath path) {
+  result = getNodeFromInputOutputPath(baseNode, path, path.getNumToken())
+}
+
+/**
+ * An access part token such as `Argument[1]` or `ReturnValue`, appearing in one or more access paths.
+ */
+class AccessPathToken extends string {
+  AccessPathToken() { this = any(AccessPath path).getRawToken(_) }
+
+  /** Gets the name of the token, such as `Member` from `Member[x]` */
+  string getName() { result = this.regexpCapture("(.+?)(?:\\[.*?\\])?", 1) }
+
+  /**
+   * Gets the argument list, such as `1,2` from `Member[1,2]`,
+   * or has no result if there are no arguments.
+   */
+  string getArgumentList() { result = this.regexpCapture(".+?\\[(.*?)\\]", 1) }
```

Comment:
Sure, I personally find the lazy operator to be more readable, but I suppose it comes with a runtime cost.

The regexp for `getArgumentList` was originally written like that to make it clear that it has no result when the brackets are omitted, but we have a comment saying this explicitly so that's probably fine.

Reply:
> but I suppose it comes with a runtime cost. 

I think the performance cost is negligible. I just wanted to have a single common regexp.